### PR TITLE
Ploop graphdriver v2

### PR DESCRIPTION
**Pull request by kolyshkin**
This is a second attempt in merging ploop graphdriver.
This time I have addressed all concerns voiced in previous iteration (https://github.com/docker/docker/pull/15516). Notable changes are:
- this no longer depends on any libraries
- vendoring commit is separate
- some more testing

The driver itself hasn't changed much since the last pull request; instead I rewrote the goploop wrapper to call ploop command line tool. I did some benchmarks and figured out performance doesn't differ much, as fork/exec overhead is negligible for most operations. 

Also, I no longer consider this experimental, it performs well and survives all the tests I threw at it (more on that in [README.md](https://github.com/kolyshkin/docker/blob/ploop-merge-2/daemon/graphdriver/ploop/README.md); if there are some more tests I should run, please let me know).

For more details, please see:
- [daemon/graphdriver/ploop/README.md](https://github.com/kolyshkin/docker/blob/ploop-merge-2/daemon/graphdriver/ploop/README.md)
- [github.com/kolyshkin/goploop-cli/blob/master/README.md](https://github.com/kolyshkin/goploop-cli/blob/master/README.md)
- [openvz.org/Ploop](https://openvz.org/Ploop)

### CODE REVIEW

**Comment by thaJeztah** [comment]
Builds failed, because there are some linting errors;

```
---> Making bundle: validate-lint (in bundles/1.9.0-dev/validate-lint)
pkg/archive/example_changes.go is in package main, not archive
Errors from golint:
daemon/daemon_ploop.go:6:2: a blank import should be only in a main or test pack
```

**Comment by thaJeztah** [comment]
ping @LK4D4 @unclejack for review

Thanks @kolyshkin, looks interesting!

**Comment by kolyshkin** [comment]
@thaJeztah 
> Builds failed, because there are some linting errors:
> pkg/archive/example_changes.go is in package main, not archive

That's not my code

> Errors from golint:
> daemon/daemon_ploop.go:6:2: a blank import should be only in a main or test pack

Thanks, fixed now, patch updated.

**Comment by kolyshkin** [comment]
Driver updated to return proper error on initializatin (graphdriver.ErrPrerequisites)

**Comment by kolyshkin** [comment]
ping @LK4D4 @unclejack please

File: daemon/graphdriver/ploop/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+This document describes how to run the Docker ploop graphdriver.
+
+Ploop is the enhanced block loop device, developed for
+[OpenVZ](https://openvz.org/) since around 2008 and used in production
+since 2012. To know more about ploop, see
+[openvz.org/Ploop](https://openvz.org/Ploop).
+
+This driver is built upon the following stack of software:
+* kernel ploop driver: [github.com/OpenVZ/vzkernel/drivers/block/ploop](https://github.com/OpenVZ/vzkernel/tree/branch-rh7-3.10.0-123.1.2-ovz/drivers/block/ploop)
```
Comment (LK4D4) [inline daemon/graphdriver/ploop/README.md]: This link is 404

Comment (LK4D4) [inline daemon/graphdriver/ploop/README.md]: Link `github.com/OpenVZ/vzkernel/drivers/block/ploop` is 404

**Comment by LK4D4** [comment]
@kolyshkin ploop modules works only with RHEL7 kernel?

**Comment by kolyshkin** [comment]
> Link github.com/OpenVZ/vzkernel/drivers/block/ploop is 404

This is not link, this is the anchor text. The link is in round brackets, and it's
https://github.com/OpenVZ/vzkernel/tree/branch-rh7-3.10.0-123.1.2-ovz/drivers/block/ploop

Although I just realized I gave a link to older kernel branch. Current one is
https://github.com/OpenVZ/vzkernel/tree/branch-rh7-3.10.0-229.7.2-ovz/drivers/block/ploop,
will update.

> ploop modules works only with RHEL7 kernel?

With RHEL6 and RHEL7 (but RHEL6 is too old to be used for Docker). We will also make them available for the recent upstream kernel (or any other kernel of your choice -- I asked @unclejack about the preferred kernel in the old pull request but got no reply).

**Comment by LK4D4** [comment]
@kolyshkin I prefer stable from www.kernel.org if it's possible. Anchor text which looks like a link looks weird :P
Now we need to find volunteers to test it.

**Comment by unclejack** [comment]
@kolyshkin Please accept my apologies for that omission. I've left out that part by mistake.

3.14.x, 3.18.x and 4.1.x are valid targets in my opinion. I'll test on any of those kernels.

Has ploop been proposed for mainline yet?

**Comment by kolyshkin** [comment]
@unclejack we'll port ploop to these (or some of these) kernels soon (I hope next week but this depends on many factors out of my control). I'll let you know once it's done.

As for upstreaming, I am not a kernel guy so I asked around and this is what I got. We have proposed it way back, I think in 2012 during kernel LSF/MM mini-summit, and are working on it on and off since that time. Since that time a few problems in the upstream kernel were identified and fixed, but this is still ongoing. The current status is some changes in (upstream) kernel are in progress (regarding kernel direct I/O and kernel async I/O), once these are settled we'll try to resubmit.

**Comment by kolyshkin** [comment]
I guess I have to reimplement this (again) as an out-of-process graphdriver via https://github.com/docker/docker/pull/13777 (once/if it's merged), so closing this one.

**Comment by thaJeztah** [comment]
Oh, sorry you'll have to do the extra miles! Thanks so much :+1:

**Pull request closed without merging**